Flaky distributed/tests/test_worker.py::test_lifetime

Random test failure of `distributed/tests/test_worker.py::test_lifetime`

First observed in https://github.com/dask/distributed/pull/4774/checks?check_run_id=2693532608

<details/>

```
    @pytest.mark.slow
    @pytest.mark.asyncio
    async def test_lifetime(cleanup):
        async with Scheduler() as s:
            async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
                async with Client(s.address, asynchronous=True) as c:
                    futures = c.map(slowinc, range(200), delay=0.1)
                    await asyncio.sleep(1.5)
                    assert b.status != Status.running
                    await b.finished()
    
>                   assert set(b.data).issubset(a.data)  # successfully moved data over
E                   assert False
E                    +  where False = <built-in method issubset of set object at 0x7f8134ca3ba0>(Buffer<<LRU: 1176/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-bi3offm0/storage, mode="a", 0 elements>>>)
E                    +    where <built-in method issubset of set object at 0x7f8134ca3ba0> = {'slowinc-0093af3b7e8e6f16b973e417caa4f85c', 'slowinc-120b294a681d6c5fc1bb125b7dc27a53', 'slowinc-28ec16f2ca5777e76e1e...eeed2f63f7030bb5542ededd', 'slowinc-4b9f7bf1e74f50ba387496336a3ab2de', 'slowinc-4eb81a10f2358a764e70dd9443ac65a8', ...}.issubset
E                    +      where {'slowinc-0093af3b7e8e6f16b973e417caa4f85c', 'slowinc-120b294a681d6c5fc1bb125b7dc27a53', 'slowinc-28ec16f2ca5777e76e1e...eeed2f63f7030bb5542ededd', 'slowinc-4b9f7bf1e74f50ba387496336a3ab2de', 'slowinc-4eb81a10f2358a764e70dd9443ac65a8', ...} = set(Buffer<<LRU: 588/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-5762snfd/storage, mode="a", 0 elements>>>)
E                    +        where Buffer<<LRU: 588/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-5762snfd/storage, mode="a", 0 elements>>> = <Worker: 'tcp://10.212.11.232:52856', tcp://10.212.11.232:52856, Status.closed, stored: 21, running: 3/3, ready: 76, comm: 0, waiting: 0>.data
E                    +    and   Buffer<<LRU: 1176/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-bi3offm0/storage, mode="a", 0 elements>>> = <Worker: 'tcp://10.212.11.232:52850', tcp://10.212.11.232:52850, Status.running, stored: 42, running: 3/3, ready: 155, comm: 0, waiting: 0>.data

distributed/tests/test_worker.py:1588: AssertionError
```

--- a/distributed/tests/test_worker.py
+++ b/distributed/tests/test_worker.py
@@ -1661,18 +1661,34 @@ async def test_close_gracefully(c, s, a, b):
 
 
 @pytest.mark.slow
-@gen_cluster(client=True, nthreads=[])
-async def test_lifetime(c, s):
-    async with Worker(s.address) as a, Worker(s.address, lifetime="2 seconds") as b:
+@gen_cluster(client=True, nthreads=[("", 1)], timeout=10)
+async def test_lifetime(c, s, a):
+    async with Worker(s.address, lifetime="2 seconds") as b:
         futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
         await asyncio.sleep(1)
         assert not a.data
-        assert b.data
-        b_keys = set(b.data)
-        while b.status == Status.running:
+        # Note: keys will appear in b.data several milliseconds before they switch to
+        # status=memory in s.tasks. It's important to sample the in-memory keys from the
+        # scheduler side, because those that the scheduler thinks are still processing
+        # won't be replicated by retire_workers().
+        mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
+        assert mem
+
+        while b.status != Status.closed:
             await asyncio.sleep(0.01)
-        await b.finished()
-        assert b_keys.issubset(a.data)  # successfully moved data over from b to a
+
+    # All tasks that were in memory in b have been copied over to a;
+    # they have not been recomputed
+    for key in mem:
+        assert_worker_story(
+            a.story(key),
+            [
+                (key, "put-in-memory"),
+                (key, "receive-from-scatter"),
+            ],
+            strict=True,
+        )
+        assert key in a.data
 
 
 @gen_cluster(worker_kwargs={"lifetime": "10s", "lifetime_stagger": "2s"})

--- a/distributed/tests/test_worker.py
+++ b/distributed/tests/test_worker.py
@@ -1635,11 +1635,9 @@ async def test_close_gracefully(c, s, a, b):
     # be replicated by retire_workers().
     while True:
         mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
-        if len(mem) >= 8:
+        if len(mem) >= 8 and any(ts.state == "executing" for ts in b.tasks.values()):
             break
         await asyncio.sleep(0.01)
-
-    assert any(ts for ts in b.tasks.values() if ts.state == "executing")
 
     await b.close_gracefully()
 
@@ -1663,16 +1661,22 @@ async def test_close_gracefully(c, s, a, b):
 @pytest.mark.slow
 @gen_cluster(client=True, nthreads=[("", 1)], timeout=10)
 async def test_lifetime(c, s, a):
+    # Note: test was occasionally failing with lifetime="1 seconds"
     async with Worker(s.address, lifetime="2 seconds") as b:
         futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
-        await asyncio.sleep(1)
-        assert not a.data
+
         # Note: keys will appear in b.data several milliseconds before they switch to
         # status=memory in s.tasks. It's important to sample the in-memory keys from the
         # scheduler side, because those that the scheduler thinks are still processing
         # won't be replicated by retire_workers().
-        mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
-        assert mem
+        while True:
+            mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
+            if len(mem) >= 8:
+                break
+            await asyncio.sleep(0.01)
+
+        assert b.status == Status.running
+        assert not a.data
 
         while b.status != Status.closed:
             await asyncio.sleep(0.01)

--- a/distributed/tests/test_worker.py
+++ b/distributed/tests/test_worker.py
@@ -1663,9 +1663,9 @@ async def test_close_gracefully(c, s, a, b):
 @pytest.mark.slow
 @gen_cluster(client=True, nthreads=[])
 async def test_lifetime(c, s):
-    async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
+    async with Worker(s.address) as a, Worker(s.address, lifetime="2 seconds") as b:
         futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
-        await asyncio.sleep(0.5)
+        await asyncio.sleep(1)
         assert not a.data
         assert b.data
         b_keys = set(b.data)

--- a/distributed/tests/test_worker.py
+++ b/distributed/tests/test_worker.py
@@ -1627,32 +1627,52 @@ async def test_worker_listens_on_same_interface_by_default(cleanup, Worker):
 
 @gen_cluster(client=True)
 async def test_close_gracefully(c, s, a, b):
-    futures = c.map(slowinc, range(200), delay=0.1)
+    futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
 
-    while not b.data:
+    # Note: keys will appear in b.data several milliseconds before they switch to
+    # status=memory in s.tasks. It's important to sample the in-memory keys from the
+    # scheduler side, because those that the scheduler thinks are still processing won't
+    # be replicated by the Active Memory Manager.
+    while True:
+        mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
+        if len(mem) >= 8:
+            break
         await asyncio.sleep(0.01)
-    mem = set(b.data)
-    proc = {ts for ts in b.tasks.values() if ts.state == "executing"}
-    assert proc
+
+    assert any(ts for ts in b.tasks.values() if ts.state == "executing")
 
     await b.close_gracefully()
 
     assert b.status == Status.closed
     assert b.address not in s.workers
-    assert mem.issubset(a.data.keys())
-    for ts in proc:
-        assert ts.state in ("executing", "memory")
+
+    # All tasks that were in memory in b have been copied over to a;
+    # they have not been recomputed
+    for key in mem:
+        assert_worker_story(
+            a.story(key),
+            [
+                (key, "put-in-memory"),
+                (key, "receive-from-scatter"),
+            ],
+            strict=True,
+        )
+        assert key in a.data
 
 
 @pytest.mark.slow
 @gen_cluster(client=True, nthreads=[])
 async def test_lifetime(c, s):
     async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
-        futures = c.map(slowinc, range(200), delay=0.1, worker=[b.address])
-        await asyncio.sleep(1.5)
-        assert b.status not in (Status.running, Status.paused)
+        futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
+        await asyncio.sleep(0.5)
+        assert not a.data
+        assert b.data
+        b_keys = set(b.data)
+        while b.status == Status.running:
+            await asyncio.sleep(0.01)
         await b.finished()
-        assert set(b.data) == set(a.data)  # successfully moved data over
+        assert b_keys.issubset(a.data)  # successfully moved data over from b to a
 
 
 @gen_cluster(worker_kwargs={"lifetime": "10s", "lifetime_stagger": "2s"})

--- a/distributed/tests/test_worker.py
+++ b/distributed/tests/test_worker.py
@@ -1632,7 +1632,7 @@ async def test_close_gracefully(c, s, a, b):
     # Note: keys will appear in b.data several milliseconds before they switch to
     # status=memory in s.tasks. It's important to sample the in-memory keys from the
     # scheduler side, because those that the scheduler thinks are still processing won't
-    # be replicated by the Active Memory Manager.
+    # be replicated by retire_workers().
     while True:
         mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
         if len(mem) >= 8:

